File: Arduino/libraries/Radio/Radio.cpp
/** ***************************************************************************
 * @file
 *
 * @brief   The implementation of the radio class
 *****************************************************************************/

#include "Radio.h"

 /** ***************************************************************************
  * @author Micah Healy
  *
  * @par Description:
  * sets up the radio class. initialize the the setup array
  *
  *****************************************************************************/
Radio::Radio()
{
    //initilise setup
    for (int i = 0; i < 6; i++)
    {
        input_setup[i] = -1;
    }
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description: 
 * sets up the radio class. initialize the the setup array 
 *
 * @param[in] setup[][2] - a 2D array containing pins to activate and the pins
 * to assign to them 
 * @param[in] size - the size to the first column
 * 
 *****************************************************************************/
Radio::Radio(int setup[][2], int size)
{
    //turn all pins to off state
    for (int i = 0; i < 6; i++)
        input_setup[i] = -1;

    //initilise setup
    for (int i = 0; i < size; i++)
    {
        int pin = setup[i][1];
        int channel = setup[i][0];
        //error checking
        if (pin > -1 && pin < 14)
        {
            //assignment
            input_setup[channel] = pin;
            pinMode(pin, INPUT);
        }
    }
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Checks inputs
 *
 * @param[in] channel - the channel to activate
 * @param[in] pin - the pin to assign to the channel
 *
 * @returns 0 - no error
 * @returns -1 - invalid pin
 * @returns -2 - invalid channel
 *
 *****************************************************************************/
int Radio::set(channel channel, int pin)
{
    //error checking
    if (pin < -1 || pin > 13)
        return -1;
    if (channel < 0 || channel > 5)
        return - 2;

    //setup pin
    if (pin != -1)
        pinMode(pin, INPUT);

    //asign in setup
    input_setup[channel] = pin;

    return 0;
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Using a given channel finds the pin number and reads the input using one of 
 * the three read functions
 *
 * @param[in] channel - the channel to read
 *
 * @returns The output from one of the read functions
 * @returns -100000 - channel has no pin set to it
 * @returns -100001 - channel values dosn't exist
 *
 *****************************************************************************/
int Radio::read(channel channel)
{
    int pin = input_setup[channel];

    if (pin == -1)
        return -100000;

    //determine channel type & read in data
    if (channel == Throttle)
        return (readThrottle(pin));
    else if (channel < Gear)
        return readJoystick(pin);
    else if (channel < 6) //switches
        return readSwitch(pin);
    else
        return -100001;
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Changes the scale that the joystick/throttle values are scale off.
 *
 * @param[in] new_scale - the scale to scale joystick/throttle values to
 *
 *****************************************************************************/
void Radio::setScale(int new_scale)
{
    scale = new_scale;
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Using a pin location reads in the duration of the pulse and outputs the
 * duration scaled from 0 to a set scale int the class
 *
 * @param[in] pin - the location that the reciver is connected
 *
 * @returns scale - a value between 0 and the scale representing the throttle 
 * position
 *
 *****************************************************************************/
int Radio::readThrottle(int pin)
{
    //read in data
    int duration = pulseIn(pin, HIGH);

    //translate to consistent scale
    int output = map(duration, 973, 1915, 0, scale);

    if (output > scale)
        return scale;
    return output;
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Using a pin location reads in the duration of the pulse and outputs the
 * duration scaled from -255 to 255 
 *
 * @param[in] pin - the location that the reciver is connected 
 *
 * @returns duration(mapped) - a value between the scale and -scale which 
 * represents the joystick position
 *
 *****************************************************************************/
int Radio::readJoystick(int pin)
{
    //read in data
    int duration = pulseIn(pin, HIGH);

    //translate to consistent scale
    return map(duration, low, high, -scale, scale);
}

/** ***************************************************************************
 * @author Micah Healy
 *
 * @par Description:
 * Using a pin number reads in the duration of the pulse and outputs the 
 * position the switch is in.
 *
 * @param[in] pin - the location that the reciver is connected 
 *
 * @returns 2 - low position
 * @returns 1 - mid position
 * @returns 0 - high position
 * @returns -256 - error
 *
 *****************************************************************************/
int Radio::readSwitch(int pin)
{
    //read in data
    int output = pulseIn(pin, HIGH);

    //translate to consistent scale
    if (abs(output - low) < 100)
        return 2;
    else if (abs(output - zero) < 100)
        return 1;
    else if (abs(output - high) < 100)
        return 0;
    else
        return -256;
}

File: Codebase_1-0/Python/Raspberry/Radio/radio.py
from enum import Enum
import RPi.GPIO as GPIO
import time

class radio:
    low = 1000
    zero = 1400
    high = 1800
    scale = 255
    input_setup

    class ports(Enum):
        Throttle = 1
        Aileron = 2
        Elevator = 3
        Rudder = 4
        Gear = 5
        AUX1 = 6
        scale = 255

    def __init__(self):

    #internal function
    def pulseIn(self, pin):
        start_time = time()
        while not GPIO.input(pin):
            start_time = time()
        while GPIO.input(pin):
            end_time = time()
        return end_time - start_time

    #internal function
    def map (self, num, low_in, high_in, low_out, high_out):
        return num * (high_out - low_out) / (high_in - low_in)

    #used to set ports
    def set(self, channel, pin):
        GPIO.setup(pin, GPIO.IN)
        self.input_setup[channel] = pin

    #used to read channels
    def read (self, channel):
        if channel == self.ports.Throttle:
           return self.readThrottle(self.input_setup[channel])
        elif channel < self.ports.Gear:
            return self.readJoystick(self.input_setup[channel])
        else:
            return self.readSwitch(self.input_setup[channel])

    #internal function
    def readThrottle(self, pin):
        duration = self.pulseIn(pin)
        output = self.map(duration, 973, 1915, 0, self.scale)
        if output > scale:
            return scale
        return output

    #internal function
    def readJoystick(self, pin):
        duration = self.pulseIn(pin)
        output = self.map(duration, 973, 1915, 0, self.scale)
        return  output

    #internal function
    def readSwitch(self, pin):
        output = self.pulseIn(pin)

        if abs(output - self.low) < 100:
            return 2
        elif abs(output - self.zero) < 100:
            return 1
        elif abs(output - self.high) < 100:
            return 0
        else:
            return -256